master/cache.lua

```diff
@@ -0,0 +1,465 @@
+          goto continue
+       end
+
+       local tasks = app["tasks"]
+
+       -- Process only tasks in TASK_RUNNING state.
+       -- From http://lua-users.org/wiki/TablesTutorial: "inside a pairs loop,
+       -- it's safe to reassign existing keys or remove them"
+       for i, t in ipairs(tasks) do
+          if t["state"] ~= "TASK_RUNNING" then
+             table.remove(tasks, i)
+          end
+       end
+
+       -- next() returns nil if table is empty.
+       local i, task = next(tasks)
+       if i == nil then
+          ngx.log(ngx.NOTICE, "No task in state TASK_RUNNING for app '" .. appId .. "'")
+          goto continue
+       end
+
+       ngx.log(
+          ngx.NOTICE,
+          "Reading state for appId '" .. appId .. "' from task with id '" .. task["id"] .. "'"
+          )
+
+       local host = task["host"]
+       if not host then
+          ngx.log(ngx.NOTICE, "Cannot find host for app '" .. appId .. "'")
+          goto continue
+       end
+
+       local ports = task["ports"]
+       if not ports then
+          ngx.log(ngx.NOTICE, "Cannot find ports for app '" .. appId .. "'")
+          goto continue
+       end
+
+       local port = ports[portIdx]
+       if not port then
+          ngx.log(ngx.NOTICE, "Cannot find port at port index '" .. portIdx .. "' for app '" .. appId .. "'")
+          goto continue
+       end
+
+       local url = scheme .. "://" .. host .. ":" .. port
+       svcApps[svcId] = {scheme=scheme, url=url}
+
+       ::continue::
+    end
+
+    svcApps_json = cjson_safe.encode(svcApps)
+
+    ngx.log(ngx.DEBUG, "Storing Marathon apps data to SHM.")
+    if not cache_data("svcapps", svcApps_json) then
+        ngx.log(ngx.ERR, "Storing marathon apps cache failed")
+        return
+    end
+
+    ngx.update_time()
+    local time_now = ngx.now()
+    if cache_data("svcapps_last_refresh", time_now) then
+        ngx.log(ngx.INFO, "Marathon apps cache has been successfully updated")
+    end
+
+    return
+end
+
+
+local function fetch_and_store_marathon_leader()
+    -- Fetch Marathon leader address. If successful, store to SHM cache.
+    -- Expected to run within lock context.
+    local mleaderRes, err = request("127.0.0.1", 8080, "/v2/leader", true)
+
+    if err then
+        ngx.log(ngx.NOTICE, "Marathon leader request failed: " .. err)
+        return
+    end
+
+    -- We need to translate 404 reply into a JSON that is easy to process for
+    -- endpoints:
+    -- https://mesosphere.github.io/marathon/docs/rest-api.html#get-v2-leader
+    local res_body
+    if mleaderRes.status == 404 then
+        -- Just a hack in order to avoid using gotos - create a substitute JSON
+        -- that can be parsed and processed by normal execution path and at the
+        -- same time passes the information of a missing Marathon leader.
+        ngx.log(ngx.NOTICE, "Using empty Marathon leader JSON")
+        res_body = '{"leader": "not elected:0"}'
```

Comment:
Either this or more code/complexity due to passing extra information to calling function :P